### PR TITLE
Fix fuerte TLS connection breakages.

### PR DESCRIPTION
**Pull request by neunhoef**
This ensures that a connection which has already been leased out of the
connection pool can no longer run into its idle timeout.

This has to be ported to 3.7 but not to older versions.

### CODE REVIEW

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10959/

**Comment by neunhoef** [comment]
This still has a 2ms idle timeout. Will fix once Jenkins is green with this.

File: 3rdParty/fuerte/src/H1Connection.h [hunk 1]
```diff
@@ -146,6 +151,11 @@ class H1Connection final : public fuerte::GeneralConnection<ST> {
   // We use this to compute the timeout of the corresponding read if the
   // write was done.
 
+  std::atomic<int>
+      _leased;  // set to 1 by the `lease` method, prevents idle alarm if
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/H1Connection.h:155]: `_leased` is nowhere initialized. We should initialize it to `0` in the constructor.

File: 3rdParty/fuerte/src/H1Connection.h [hunk 2]
```diff
@@ -146,6 +151,11 @@ class H1Connection final : public fuerte::GeneralConnection<ST> {
...
+                // non-zero, set to 2 by `sendRequest` method, set back to 0
+                // when the idle alarm is set, provided it was 2
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/H1Connection.h:157]: Also: `it is temporarily set to -1 during shutdown of the connection`.

File: arangod/Network/ConnectionPool.cpp [hunk 3]
```diff
@@ -177,7 +177,8 @@ ConnectionPool::Context::Context(std::shared_ptr<fuerte::Connection> c,
     : fuerte(std::move(c)), lastLeased(t), leases(l) {}
 
 std::shared_ptr<fuerte::Connection> ConnectionPool::createConnection(fuerte::ConnectionBuilder& builder) {
-  auto idle = std::chrono::milliseconds(_config.idleConnectionMilli);
+  //auto idle = std::chrono::milliseconds(_config.idleConnectionMilli);
+  auto idle = std::chrono::milliseconds(2);
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp:181]: Fine for testing, but we need remember to revert this change before merging.

**Comment by graetzer** [comment]
Why don‘t you all just remove the idle timeout instead of messing with it

**Comment by neunhoef** [comment]
Because then connections stay open indefinitely. Probably idle timeout should be a feature of the connection pool and not of the connection object itself. But this is a larger change. Finally, I would not call it "messing with it" but "repairing it"...

**Comment by neunhoef** [comment]
I will port the additional comments from the 3.6 version tomorrow to this branch.

**Comment by graetzer** [comment]
The connection pool will remove unused connections

**Comment by graetzer** [comment]
I also like how now there are 3 connection usage counters

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10981/

File: 3rdParty/fuerte/src/H1Connection.cpp [hunk 4]
```diff
@@ -619,15 +648,24 @@ void H1Connection<ST>::setTimeout(std::chrono::milliseconds millis, TimeoutType
...
-            me->shutdownConnection(Error::CloseRequested);
+            int exp = 0;
+            if (me->_leased.compare_exchange_strong(exp, -1)) {
+              // Note that we check _leased here with std::memory_order_seq_cst
+              // **before** we call `shutdownConnection` and actually kill the
+              // connection. This is important, see below!
+              FUERTE_LOG_DEBUG << "HTTP-Request idle timeout"
+                               << " this=" << me << "\n";
+              me->shutdownConnection(Error::CloseRequested);
+              me->_leased = 0;
```
Comment (mpoeter) [inline 3rdParty/fuerte/src/H1Connection.cpp:667]: > Note that we check _leased here with std::memory_order_seq_cst **before** we call `shutdownConnection` [...]

Actually, it is more important that this assignment to `_leased` uses at least `std::memory_order_release`. This is necessary to ensure that the compare_exchange in `lease` synchronizes-with this store, thereby ensuring that the subsequent check of `_state` in `lease` sees the value written in `shutdownConnection` (regardless which memory ordering was used there).

That's why I usually prefer store/load calls with explicit memory ordering instead of the (more convenient) operator overloads.
```suggestion
              me->_leased.store(0, std::memory_order_seq_cst);
```

File: arangod/Network/ConnectionPool.cpp (hunk 3 above)
Comment (neunhoef) [reply arangod/Network/ConnectionPool.cpp:181]: Indeed!

File: 3rdParty/fuerte/src/H1Connection.cpp (hunk 4 above)
Comment (neunhoef) [reply 3rdParty/fuerte/src/H1Connection.cpp:667]: I am happy with this change.

**Comment by neunhoef** [comment]
The final Jenkins run with normal idle timeout will test this latest commit, so no need to stop the already running jenkins.

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10991/

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11001/

File: 3rdParty/fuerte/src/H1Connection.h (hunk 1 above)
Comment (neunhoef) [reply 3rdParty/fuerte/src/H1Connection.h]: Done.

File: 3rdParty/fuerte/src/H1Connection.cpp [hunk 5]
```diff
@@ -405,6 +411,8 @@ void H1Connection<ST>::asyncWriteNextRequest() {
       if (_shouldKeepAlive && this->_config._idleTimeout.count() > 0) {
         FUERTE_LOG_HTTPTRACE << "setting idle keep alive timer, this=" << this
                              << "\n";
+        int exp = 2;
+        _leased.compare_exchange_strong(exp, 0);
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply 3rdParty/fuerte/src/H1Connection.cpp]: No, all 4 values can occur.

File: 3rdParty/fuerte/src/H1Connection.cpp [hunk 6]
```diff
@@ -211,6 +211,12 @@ void H1Connection<ST>::sendRequest(std::unique_ptr<Request> req,
 
   FUERTE_LOG_HTTPTRACE << "queued item: this=" << this << "\n";
 
+  // If the connection was leased explicitly, we set the _leased indicator
+  // to 2 here, such that it can be reset to 0 once the idle alarm is
+  // enabled explicitly later
+  int exp = 1;
+  _leased.compare_exchange_strong(exp, 2);
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply 3rdParty/fuerte/src/H1Connection.cpp:225]: No, all four values can occur.

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11011/ final test with normal idle timeout.

File: 3rdParty/fuerte/src/H1Connection.cpp [hunk 7]
```diff
@@ -405,6 +418,20 @@ void H1Connection<ST>::asyncWriteNextRequest() {
...
+        // 0, if _leased is still 1 (`lease` has happened, but `sendRequest`
+        // has not yet been called), then we leave it untouched. Since
+        // this method here is only ever executed on the iothread and
+        // the value -1 only happens for a brief period in the iothread,
+        // it will never be observed here. Therefore, no actual need to check
+        // the result of the compare_exchange_strong.
+        int exp = 2;
+        if (!_leased.compare_exchange_strong(exp, 0)) {
+          FUERTE_ASSERT(exp != -1);
+        };
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/H1Connection.cpp:434]:
```suggestion
        }
```

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11035/

**Comment by neunhoef** [comment]
One more run after merging in latest changes.
